---
title: "CERES for R"
author: "Jordan Bryan"
date: "2017-07-27"
output:
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CERES for R}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

CERES is an algorithm that estimates gene essentiality from sgRNA depletion data while accounting for:

1. Effects of Cas9-mediated DNA cleavage
2. Differential sgRNA on-target activity

Running CERES requires copy number data for all targeted genomic loci as well as mappings from genomic loci to gene symbols.

## Installation

To install `ceresr`, first download the linked `.tar` file from the `R package` link on the reviewer site. Preparing common file formats for CERES input depends on several packages from [Bioconductor](http://bioconductor.org). To install these dependencies, begin an R session and execute the following:

```
source("https://bioconductor.org/biocLite.R")
biocLite(

  pkgs=c("Biostrings", "Rsamtools", "GenomeInfoDb", 
         "BSgenome", "BSgenome.Hsapiens.UCSC.hg19", 
         "GenomicRanges")
         
  )
```

Once the packages have completed installation, execute the following from the command line:

```
R CMD INSTALL ceresr_0.0.0.9000.tar
```

## Examples

The `wrap_ceres` function provides the main interface to the CERES algorithm. You can run CERES on the provided input files with:

```
ceres_results <- wrap_ceres(sg_data = "guide_sample_dep.Rds", 
                            cn_data = "locus_sample_cn.Rds", 
                            guide_locus = "guide_locus_map3.Rds", 
                            locus_gene = "locus_gene_map3.Rds", 
                            replicate_map = "replicate_map.Rds", 
                            run_id = "my_crispr_screen", 
                            optimize = F)
```

## Algorithm Description

CERES infers gene effects and copy number effects via alternating constrained least squares, solving the problem:

\begin{equation*}
\begin{aligned}
& \underset{\hat{D}}{\text{minimize}}
& & \sum_{i = 1}^M \sum_{j = 1}^N (D_{ij} - \hat{D}_{ij})^2 + \lambda_g\sum_{j=1}^N\sum_{k=1}^K g_{kj}^2\\ \\
& \text{subject to}
& & 0 \leq q_i \leq 1, \; \; i = 1, \ldots, M \\
& & & f_j(C) \leq f_j(C'), \; \; \forall {C \geq C'} \in \mathbb{R}_{\geq 0}, \; \; j = 1, \dots , N
\end{aligned}
\end{equation*}

where

\begin{equation*}
\begin{aligned}
\hat{D}_{ij} = q_i\left(\sum_{k \in \mathcal{G}_i} ({h}_{k} + g_{kj}) + f_j\left(\textstyle\sum_{l \in \mathcal{C}_i}C_{lj}\right)\right) + o_i
\end{aligned}
\end{equation*}

### CERES alternating minimization 

| **0.** given $\epsilon > 0$, $[g, f] := 0$, $[q, o] := [1, 0]$;
|
| Repeat:
|
|   **1.** *Solve for gene and copy-number effects*. Compute optimal parameters $[g^*, f^*]$
|   **2.** Update. $[g, f] := [g^*, f^*]$
|   **3.** *Solve for sgRNA efficacies and offsets*. Compute optimal parameters $[q^*, o^*]$
|   **4.** Update. $[q, o] := [q^*, o^*]$
|   **5.** Evaluate error. $\text{mse}_{t} := \frac{||D - \hat{D}||^2}{MN}$
|   **6.** Evaluate decrease in error. $\Delta \text{mse} := \text{mse}_{t} - \text{mse}_{t-1}$
|   **7.** Stopping criterion. quit if $\Delta \text{mse} < \epsilon$